docs/developer/architecture/configurations.rst

```diff
@@ -73,3 +73,197 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+fully ready and available we could perform the following actions:
+
+  - Create and deploy ConfigMaps that hold cluster and service configurations
+    and pre-fill them with default values.
+  - Template service pods and deployments to consume configuration data
+    directly from the above deployed ConfigMaps
+
+This approach works because in a MetalK8s cluster, ConfigMaps for cluster and
+service configurations are available before we deploy the configured services.
+
+**Using Salt states**
+
+Once a ConfigMap is updated by the user (say a user changes the number of
+replicas for Prometheus deployments to a new value), then perform the
+following actions:
+
+  - Apply a Salt states that reads the ConfigMap object, validates the schema
+    and checks the new values passed.
+
+Storage format
+~~~~~~~~~~~~~~
+
+A YAML (K8s-like) format was chosen to represent the data field instead of a
+flat key-value structure for the following reasons:
+
+ - YAML formatted configurations are easy to write and understand hence it will
+   be simpler for users to edit configurations.
+ - The YAML format benefits from bearing a schema version, which can be checked
+   and validate against a version we deploy.
+ - YAML is a format for describing hierarchical data structures, while using a
+   flat key-value format would require a form of encoding (and then, decoding)
+   of this hierarchical structure.
+
+A sample ConfigMap can be defined with the following fields.
+
+An example of such a ConfigMap:
+
+.. code-block:: yaml
+
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      namespace: <namespace>
+      name: <config-name>
+    data:
+      config.yaml: |-
+        apiVersion: <object-version>
+        kind: <kind>
+        spec:
+          <key>: <values>
+
+**Use case 1:**
+
+Configure and store the number of replicas for service specific Deployments
+found in the `metalk8s-monitoring` namespace using the ConfigMap format.
+
+.. code-block:: yaml
+
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      namespace: metalk8s-monitoring
+      name: metalk8s-grafana-config
+    data:
+      config.yaml: |-
+        apiVersion: metalk8s.scality.com/v1alpha1
+        kind: GrafanaConfig
+        spec:
+          deployment:
+            replicas: 2
+
+Non-goals
+~~~~~~~~~
+
+This section contains requirements stated above which the current design choice
+does not cater for and will be addressed later:
+
+- Persisting newly added Grafana dashboards or new Grafana datasources
+  especially for modifications added via the Grafana UI cannot be stored in
+  ConfigMaps and hence will be catered for later.
+
+- As stated in the requirements, adding and editing Prometheus alert rules
+  is also not covered by the chosen design choice and will need to be addressed
+  differently. Even if we could use ConfigMaps for Prometheus rules, we prefer
+  relying on the Prometheus Operator and it's CRD (PrometheusRule).
```

Comment:
Alertmanager notification configurations are handled in our current design since we can store whatever notification configuration ( e.g slack, email) we wish to implement in ConfigMaps. PS We have a ticket for storing the Alertmanager notification configurations.

For changing those thresholds, we found out that you can actually do it in the original PrometheusRule CRD deployed by the Prometheus-operator and I want to believe the Crds are persisted even after upgrades.
We just need to provide adequate documentation explaining how this can be done.

Remember that this design doc is about using ConfigMaps to store and persist Service & Cluster configurations. Persisting or changing Prometheus rules is not handled using ConfigMaps and hence we were advised not to add that particular solution in this document.

Reply:
> do it in the original PrometheusRule CRD deployed by the Prometheus-operator and I want to believe the Crds are persisted even after upgrades.

CRDs belong to the Operator - a cluster admin could add rules by creating `PrometheusRule` objects, as he would add volumes by creating `Volume` objects.
Those *new* objects are not known by us, so we will leave them untouched under all circumstances.

> we were advised not to add that particular solution in this document.

At least in this first design PR. It may make sense to group all possible solutions to the general "configuration" problem under a single umbrella, but for now, we focus on what's missing: said ConfigMaps.